### DISCUSSION TITLE
idiomatic way to trigger file requests after dial is complete

### DISCUSSION THREAD

**Discussion by snowmead** [libp2p/rust-libp2p#4910] [Q&A] [answered]
What is the libp2p idiomatic way to trigger additional requests (in this case, file requests) after successfully dialing another node?

Right now I simply sleep for 5 seconds and it works, but I would like to trigger this based on the `ConnectionEstablished` Swarm event. 

https://github.com/snowmead/storage-provider-node/blob/main/src/main.rs#L230-L232

Should I move this [call](https://github.com/snowmead/storage-provider-node/blob/main/src/main.rs#L234-L240) to happen inside the [`ConnectionEstablised` event](https://github.com/snowmead/storage-provider-node/blob/main/src/main.rs#L168-L176). But in order to know in this context if I should trigger a file request, I would have to add additional book keeping data here or part of the `EventLoop` struct to track which peers to request a file from.

**Comment by thomaseizinger** [comment]
`libp2p-request-response` does have the ability to automatically dial a node in order to send a request if you don't already have a connection.

**Reply by snowmead** [reply]
But in my case, the peer is unknown to the node and `request_response` accepts only a `PeerId` as a parameter for `RequestFile`. I pass in any multiaddr (expecting this protocol "/ip4/.../tcp/.../p2p/...") and want to dial and request a file from it. Hence why the current dial and request is working because I dial the entire multiaddr, not just the peerId. Do I just have to modify the request_response protocol to support this for my case? I ask because I want to do things right and not assume things.

**Reply by thomaseizinger** [reply, marked as answer]
> I pass in any multiaddr (expecting this protocol "/ip4/.../tcp/.../p2p/...")

The part at after `/p2p` is the `PeerId`, so you can extract it from there :)

You can add this address first using `add_address` to keep it around and then send requests to the `PeerId` and `request-response` will use the (and other addresses) to make a new connection.

**Reply by snowmead** [reply]
Ok it worked like a charm! Thank you! :)

```rust
  self.swarm
      .behaviour_mut()
      .request_response
      .add_address(&peer_id, addr.clone());

  self.event_sender
      .send(Event::OutboundRequest {
          peer: peer_id,
          file_name,
      })
      .await
      .expect("Event receiver not to be dropped.");
```
